File: catalog/backends/node-express/patterns/rag_seed_mongo_route.js
const { MongoClient } = require('mongodb');
const { OpenAI } = require('openai');
const tracer = require('dd-trace');

// Pattern: LLM Obs — Embed and store documents in MongoDB for RAG
// Adapt: replace document schema with domain-specific fields
const client = new MongoClient(process.env.MONGODB_URL || 'mongodb://demo:demo@mongodb:27017/demo');
const db = client.db();
const oai = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });
const llmobs = tracer.llmobs;

const _embed = llmobs.wrap({ kind: 'embedding', modelName: 'text-embedding-3-small', modelProvider: 'openai' },
  async function embed(texts) {
    const resp = await oai.embeddings.create({ model: 'text-embedding-3-small', input: texts });
    return resp.data.map(e => e.embedding);
  }
);

async function seedDocuments(req, res) {
  const docs = req.body.documents || [];
  const texts = docs.map(d => d.content);
  const vectors = await _embed(texts);
  const records = docs.map((doc, i) => ({ title: doc.title, content: doc.content, embedding: vectors[i] }));
  await db.collection('documents').insertMany(records);
  res.json({ stored: docs.length });
}

module.exports = { seedDocuments };

File: catalog/backends/python-django/patterns/rag_seed_mongo_route.py
import os
import json
from openai import OpenAI
from pymongo import MongoClient
from django.http import JsonResponse
from ddtrace.llmobs.decorators import embedding

# Pattern: LLM Obs — Embed and store documents in MongoDB for RAG
# Adapt: replace document schema with domain-specific fields
MONGO_URL = os.environ.get("MONGODB_URL", "mongodb://demo:demo@mongodb:27017/demo")
_oai = OpenAI(api_key=os.environ.get("OPENAI_API_KEY"))
_db = MongoClient(MONGO_URL).get_default_database()


@embedding(model_name="text-embedding-3-small", model_provider="openai")
def _embed(texts):
    resp = _oai.embeddings.create(model="text-embedding-3-small", input=texts)
    return [e.embedding for e in resp.data]


def seed_documents(request):
    body = json.loads(request.body)
    docs = body.get("documents", [])
    texts = [d["content"] for d in docs]
    vectors = _embed(texts)
    for doc, vec in zip(docs, vectors):
        _db.documents.insert_one({"title": doc["title"], "content": doc["content"], "embedding": vec})
    return JsonResponse({"stored": len(docs)})
